### PR TITLE
Features/guild stars to rank

### PR DESCRIPTION
**Pull request by PhoenixOrigin**
Same feature, has been fixed, @magicus @kristofbolyai can either of you please help me with adding clickable urls? Since I make a new text the click events don't remain. I am not sure how to do it and I didn't understand how it was done in the other area.

### CODE REVIEW

**Comment by kristofbolyai** [comment]
You still have changes in buils gradle and changelog.md

**Comment by PhoenixOrigin** [comment]
shit sorri

**Comment by PhoenixOrigin** [comment]
Alright bolyai i fixed

**Comment by kristofbolyai** [comment]
They are still there.

**Comment by PhoenixOrigin** [comment]
Wait what can you add a comment on what I did I just pulled it from alpha from main

**Comment by kristofbolyai** [comment]
> Wait what can you add a comment on what I did I just pulled it from alpha from main

image

**Comment by PhoenixOrigin** [comment]
> > Wait what can you add a comment on what I did I just pulled it from alpha from main
> 
> image

I have no idea how to fix that i just used intellij to update with alpha

**Comment by magicus** [comment]
@PhoenixOrigin So your goal is to replace e.g. `**` with `Captain`. but keep everything else, including color codes, intact? 

Replacing parts of TextComponents is actually much, much more difficult than it seems. Your current code seems completely wrong; you can't just append the message to each and every sibling.

**Comment by PhoenixOrigin** [comment]
> @PhoenixOrigin So your goal is to replace e.g. `**` with `Captain`. but keep everything else, including color codes, intact?
> 
> Replacing parts of TextComponents is actually much, much more difficult than it seems. Your current code seems completely wrong; you can't just append the message to each and every sibling.

Yep but I tested it and the code works? I know it seems a lot more difficult but, for example in testing .getSiblings() returns 
`msgTextComponent{text='', siblings=[TextComponent{text=' testing ', siblings=[], style=Style{ color=null, bold=null, italic=null, underlined=null, strikethrough=null, obfuscated=null, clickEvent=null, hoverEvent=null, insertion=null, font=minecraft:default}}, TextComponent{text='https://google.com/', siblings=[], style=Style{ color=null, bold=null, italic=null, underlined=null, strikethrough=null, obfuscated=null, clickEvent=ClickEvent{action=OPEN_URL, value='https://google.com/%27%7D, hoverEvent=null, insertion=null, font=minecraft:default}}, TextComponent{text=' fuck coding', siblings=[], style=Style{ color=null, bold=null, italic=null, underlined=null, strikethrough=null, obfuscated=null, clickEvent=null, hoverEvent=null, insertion=null, font=minecraft:default}}], style=Style{ color=aqua, bold=null, italic=null, underlined=null, strikethrough=null, obfuscated=null, clickEvent=null, hoverEvent=null, insertion=null, font=minecraft:default}}`

As you can see for me it didn't give the name but just the message. So I can just append it no? I don't see why not and I dont think I am capable of doing it in a different way.

File: common/src/main/java/com/wynntils/features/user/GuildStarsToRankFeature.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+import com.wynntils.core.features.UserFeature;
+import com.wynntils.wynn.event.ChatMessageReceivedEvent;
+import com.wynntils.wynn.objects.GuildRank;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import net.minecraft.network.chat.MutableComponent;
+import net.minecraft.network.chat.TextComponent;
+import net.minecraftforge.eventbus.api.SubscribeEvent;
+
+public class GuildStarsToRankFeature extends UserFeature {
```
Comment (P0keDev) [inline common/src/main/java/com/wynntils/features/user/GuildStarsToRankFeature.java:18]: IMO this feature is one that should be disabled by default via `@StartDisabled` - it's a subjective change to a core wynncraft feature

**Review by P0keDev — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 2]
```diff
@@ -0,0 +1,51 @@
+## [0.0.1-alpha.117](https://github.com/Wynntils/Artemis/compare/v0.0.1-alpha.116...v0.0.1-alpha.117) (2022-12-03)
```
Comment (kristofbolyai) [inline CHANGELOG.md]: Still here

**Comment by P0keDev** [comment]
> @PhoenixOrigin So your goal is to replace e.g. `**` with `Captain`. but keep everything else, including color codes, intact?
> 
> Replacing parts of TextComponents is actually much, much more difficult than it seems. Your current code seems completely wrong; you can't just append the message to each and every sibling.

the code doesn't append something to each sibling, it appends each sibling to the reconstructed "base" message which contains the guild rank tag

that said - @PhoenixOrigin it would be beneficial to include screenshots of the feature working in-game

File: common/src/main/java/com/wynntils/features/user/GuildStarsToRankFeature.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+    @SubscribeEvent
+    public void onChat(ChatMessageReceivedEvent event) {
+        String message = event.getOriginalCodedMessage();
+        Matcher match = guildMessage.matcher(message);
+        if (!(match.matches()) || match.group(2).equals("INFO")) return;
+        MutableComponent newMessage = new TextComponent(String.format(
+                "§b[§b%s §3%s§b] §b",
+                GuildRank.fromStars(match.group(1).length()).getRank(), match.group(2)));
+        event.getMessage().getSiblings().forEach(newMessage::append);
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/features/user/GuildStarsToRankFeature.java]: This assumes that we haven't changed the component sibling tree. But that is not true. We change it in multiple occasions in the mod

File: common/src/main/java/com/wynntils/wynn/objects/GuildRank.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        this.rank = rank;
+        this.stars = stars;
+    }
+
+    public static GuildRank fromStars(int stars) {
+        for (GuildRank type : values()) {
+            if (stars == type.getStars()) {
+                return type;
+            }
+        }
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/wynn/objects/GuildRank.java]: Can be simplified using GuildRank.values()

File: common/src/main/java/com/wynntils/wynn/objects/GuildRank.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        return GuildRank.None;
+    }
+
+    public static GuildRank fromString(String rank) {
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/wynn/objects/GuildRank.java]: Is this even used?

File: build.gradle [hunk 6]
```diff
@@ -16,7 +16,7 @@ def usingHotswapAgent = project.hasProperty('wynntils.hotswap') ? project.getPro
 
 // If we are running in a CI environment, we want to use the version else dev version
 // On the release branches, this string will be automatically updated by the bots
-version = "0.0.1-main-dev"
+version = "0.0.1-alpha.117"
```
Comment (kristofbolyai) [inline build.gradle]: Same, still here

**Comment by PhoenixOrigin** [comment]
> > @PhoenixOrigin So your goal is to replace e.g. `**` with `Captain`. but keep everything else, including color codes, intact?
> > Replacing parts of TextComponents is actually much, much more difficult than it seems. Your current code seems completely wrong; you can't just append the message to each and every sibling.
> 
> the code doesn't append something to each sibling, it appends each sibling to the reconstructed "base" message which contains the guild rank tag
> 
> that said - @PhoenixOrigin it would be beneficial to include screenshots of the feature working in-game

I shall do that later

File: common/src/main/java/com/wynntils/features/user/GuildStarsToRankFeature.java (hunk 3 above)
Comment (PhoenixOrigin) [reply common/src/main/java/com/wynntils/features/user/GuildStarsToRankFeature.java]: Code still works ?

Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/features/user/GuildStarsToRankFeature.java]: ? No it does not. Even if, for some forsaken reason, it works with chat timestamps, I won't accept code that "just works"

**Comment by kristofbolyai** [comment]
This has been stale for long enough. Either finish it, or I will close this.

**Comment by magicus** [comment]
@PhoenixOrigin I am closing this PR now. If you decide to continue working on this functionality, please open a new PR from your fork.

**Pull request closed without merging**